File: typescript/src/day_02/submarine.ts
export interface Coordinates {
  horizontal: number;
  depth: number;
}

export interface Move {
  command: string;
  value: number;
}

export const calculatePositionAndDepth = (inputs: Move[]) : Coordinates => {
  const coordinates = { horizontal: 0, depth: 0 };

  inputs.forEach(input => {
    switch (input.command) {
      case 'forward':
        coordinates.horizontal += input.value;
        break;
      case 'down':
        coordinates.depth += input.value;
        break;
      case 'up':
        coordinates.depth -= input.value;
        break;
    }
  })

  return coordinates;
}

export const advancedCalculatePositionAndDepth = (inputs: Move[]) : Coordinates => {
  const coordinates = { horizontal: 0, depth: 0 };
  let aim = 0;

  inputs.forEach(input => {
    switch (input.command) {
      case 'forward':
        coordinates.horizontal += input.value;
        coordinates.depth += aim * input.value;
        break;
      case 'down':
        aim += input.value;
        break;
      case 'up':
        aim -= input.value;
        break;
    }
  })

  return coordinates;
}

File: python/day_02/submarine.py
import os

def main():
  dirname = os.path.dirname(__file__)
  filename = os.path.join(dirname, '../../data/day_02/input.txt')
  inputs = []

  with open(filename) as f:
    lines = f.readlines()

    for line in lines:
      command, value = line.split()
      inputs.append([command, int(value)])

  coordinates = calculate_position_and_depth(inputs)
  position = coordinates['horizontal'] * coordinates['depth']

  print("Horizontal: {} | Depth: {} | Result: {}".format(coordinates['horizontal'], coordinates['depth'], position))

  print('====== CALCULATING USING ADVANCED OPERATION ==========')

  coordinates = advanced_calculate_position_and_depth(inputs)
  position = coordinates['horizontal'] * coordinates['depth']

  print("Horizontal: {} | Depth: {} | Result: {}".format(coordinates['horizontal'], coordinates['depth'], position))

def calculate_position_and_depth(inputs):
  coordinates = { 'horizontal': 0, 'depth': 0}

  for command, value in inputs:
    if command == 'forward':
      coordinates['horizontal'] += value
    elif command == 'down':
      coordinates['depth'] += value
    elif command == 'up':
      coordinates['depth'] -= value

  return coordinates

def advanced_calculate_position_and_depth(inputs):
  coordinates = { 'horizontal': 0, 'depth': 0}
  aim = 0

  for command, value in inputs:
    if command == 'forward':
      coordinates['horizontal'] += value
      coordinates['depth'] += aim * value
    elif command == 'down':
      aim += value
    elif command == 'up':
      aim -= value

  return coordinates

if __name__ == '__main__':
    main()
